src/mocap_pose.cpp

```diff
@@ -410,3 +427,57 @@ MocapPose::~MocapPose()
         impl_->worker_thread.join();
     }
 }
+
+/* connection to QTM server works the following way:
+
+    - client opens TCP control connection, on which it negotiates start of UDP stream back to client
+    - if any NAT gateways stand between QTM server and the client, from gateway perspective it looks
+      like the server opens an unsolicited connection to client (server sends first packet)
+    - therefore we send a dummy packet to server so from NAT gw perspective the server-initiated packets
+      look like packets that initiated by the client (client sent first packet, which is OK).
+
+    for extra details see https://tailscale.com/blog/how-nat-traversal-works/
+
+    an alternative would have been to use full TCP connection (data plane uses the same TCP connection
+    as the control plane), but network unreliability issues make mocap-pose stuck on TCP-only, as
+    the client doesn't realize the connection is broken because after handshake + stream starting,
+    the client doesn't send any data towards the server.
+
+    TCP keepalives could have been a solution, but the latency of this data is important and I think
+    UDP based stream (considering we only care about the newest data) recovers much faster than even
+    tuned TCP keepalive parameters could.
+
+    from tcpdump looks like this (the first line is our holepunch packet, from second on its QTM data stream):
+
+    15:47:39.883615 IP worklaptop.17535 > 172.18.32.20.22225: UDP, length 0
+    15:47:39.923224 IP 172.18.32.20.22225 > worklaptop.17535: UDP, length 712
+*/
+int MocapPose::sendNATHolepunchPacket(unsigned short receivingUDPPort) {
+    uint16_t qtmServerPort = impl_->basePort + 3; // usually 22225
```

Comment:
I don't know where the +3 comes from, but basePort is 22222 and the server-sent UDP data seems to be coming from port + 3 so I assume there's a relation

Reply:
This is the reason:
https://docs.qualisys.com/qtm-rt-protocol/#ip-port-numbers

Reply:
Thanks! I'll document it in the code